go/cmd/pgctld/command/start.go

```diff
@@ -0,0 +1,371 @@
+	if err := os.MkdirAll(dataDir, 0700); err != nil {
+		return fmt.Errorf("failed to create data directory: %w", err)
+	}
+
+	// Run initdb
+	cmd := exec.Command("initdb", "-D", dataDir, "--auth-local=trust", "--auth-host=md5")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("initdb failed: %w", err)
+	}
+
+	// Set up postgres user password for md5 authentication
+	if err := setPostgresPassword(dataDir); err != nil {
+		return fmt.Errorf("failed to set postgres password: %w", err)
+	}
+
+	return nil
+}
+
+func setPostgresPassword(dataDir string) error {
+	// Start PostgreSQL temporarily in single-user mode to set password
+	// Use postgres in single-user mode with trust auth to set the password
+	cmd := exec.Command("postgres", "--single", "-D", dataDir, "postgres")
+	cmd.Stdin = strings.NewReader("ALTER USER postgres PASSWORD 'postgres';\n")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to set postgres password: %w", err)
+	}
+
+	return nil
+}
+
+func isPostgreSQLRunning(dataDir string) bool {
+	// Check if postmaster.pid file exists and process is running
+	pidFile := filepath.Join(dataDir, "postmaster.pid")
+	if _, err := os.Stat(pidFile); err != nil {
+		return false
+	}
+
+	// Read PID from file and check if process is actually running
+	pid, err := readPostmasterPID(dataDir)
+	if err != nil {
+		return false
+	}
+
+	return isProcessRunning(pid)
+}
+
+func startPostgreSQL(dataDir string) error {
+	// Legacy function - use viper config
+	config := NewPostgresConfigFromViper()
+	config.DataDir = dataDir
+	return startPostgreSQLWithConfig(config)
+}
+
+func startPostgreSQLWithConfig(config *PostgresConfig) error {
+	// Use pg_ctl to start PostgreSQL properly as a daemon
+	args := []string{
+		"start",
+		"-D", config.DataDir,
+		"-l", filepath.Join(config.DataDir, "postgresql.log"),
+		"-W", // don't wait - we'll check readiness ourselves
+	}
+
+	slog.Info("Starting PostgreSQL with configuration", "port", config.Port, "dataDir", config.DataDir)
+
+	pgOptions := []string{
+		fmt.Sprintf("-p %d", config.Port),
+	}
+
+	if config.SocketDir != "" {
+		pgOptions = append(pgOptions, fmt.Sprintf("-k %s", config.SocketDir))
+	}
+
+	if config.ConfigFile != "" {
+		pgOptions = append(pgOptions, fmt.Sprintf("-c config_file=%s", config.ConfigFile))
+	}
+
+	if len(pgOptions) > 0 {
+		args = append(args, "-o", strings.Join(pgOptions, " "))
+	}
+
+	cmd := exec.Command("pg_ctl", args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to start PostgreSQL with pg_ctl: %w", err)
+	}
+
+	// Wait for PostgreSQL to be ready using pg_isready
+	return waitForPostgreSQLWithConfig(config)
+}
+
+// isPostgreSQLReady checks if PostgreSQL is ready to accept connections
+// by reading the postgresql.log file and checking for "ready" status
```

Comment:
it's better to use the `pg_isready` binary for this. see [here](https://www.postgresql.org/docs/current/app-pg-isready.html)

Reply:
there also a exist a function `waitForPostgreSQLWithConfig()` which uses `pg_ready`. not sure if the function using log file to check status is even being used anywhere.

Reply:
removed the log check and used pg_isready everywhere.